onedal/datatypes/table.cpp

```diff
@@ -72,6 +72,10 @@ ONEDAL_PY_INIT_MODULE(table) {
         const auto column_count = t.get_column_count();
         return py::make_tuple(row_count, column_count);
     });
+    table_obj.def_property_readonly("dtype", [](const table& t){
+        // returns a numpy dtype, even if source was not from numpy
```

Comment:
Why is this required? Will this be consistent for different array inputs?

Reply:
@samir-nasibli good question! This simplifies dtype checking for the backend handoff, what oneDAL cares about most is what the dtype of the tables to it are (i.e. the output of to_table). It is not guaranteed to match the dtype of the input into to_table, but it is guaranteed in matching the oneDAL table dtype.  It is by luck that dpnp/dpctl allow for dtype comparisons to numpy dtypes, which is not the case for array_api_strict inputs. By having a consistent numpy dtype to oneDAL tables in python, we will not have to worry about supporting all the different dtypes of all the various array inputs we will eventually support.

Reply:
@icfaust Thank you for the clarification! I don't like this extension of table API which is wo for numpy and dpnp/dpctl.

I would suggest to solve the problem at python level https://github.com/intel/scikit-learn-intelex/blob/5bb54a520f5f3d1cf719e149df6526de943e393b/onedal/utils/validation.py#L452 without needing to extend table API

Reply:
@samir-nasibli I disagree, I think we should find someone to arbitrate on this.

Reply:
Also you can name it as `numpy_dtype` instead of `dtype`, but again I don't see a lot of profit having this attr for the numeric table. Let's have a broader discussion.

Reply:
Could you clarify the situation that this would be helping? Like if an alternative array backend is used then the dtype of the input to to_table may not match the output of from_table?

Reply:
The large problem coming to us soon is handling zero-copy of arrays to the backend via the array_api standard.  Comparison to numpy-dtypes when an object has different dtypes (for example an array_api_strict float64 dtype is not equivalent to a numpy float64 dtype, and will yield a false). We will have to do one of two things:

adapt the often:
"float" if dtype==np.float64 else "double" which occurs when creating param dicts
or
always use the tables dtype, which has already been translated into a oneDAL, appropriate datatype.

By using numpy as the datatype, we are not having to re-invent the wheel with creating our own dtype standard, as pybind11 supports it natively in C++, its already all over the codebase in sklearnex/onedal.

Secondarily, I have talked about how actually our backend can ingest complex float and complex double types, but will treat them as float types or double types. See here:

https://github.com/intel/scikit-learn-intelex/pull/2172#issuecomment-2495364